### ISSUE TITLE
Envsubst script is slow to execute on low end systems

### ISSUE THREAD

**Issue by eerotki** [nginxinc/docker-nginx-unprivileged#107]
I discovered that startup script `20-envsubst-on-templates.sh` takes longer than expected to execute with the recent update, when `/etc/nginx/conf.d/` path is in read-only mode. With `1.23.0-alpine`, the container is started in 4 seconds, from creation to serving traffic. While in `1.23.1-alpine`, the same configuration with same computing resources takes up to **120 seconds** to start. The script `20-envsubst-on-templates.sh` alone took 118 seconds to process. My configuration expects the container to start within 35 seconds, which has been plenty with version prio to `1.23.1`. Adjusting `conf.d` path to read-write for `1.23.1-alpine` reduces the startup time as it was, to 4 seconds.

I'm using GKE 1.22 + Docker 20.10.6 with small resources (25m CPU, 50Mi RAM). This configuration has been working flawlessly in the past years. When ran locally with my mighty laptop the start up is pretty much instantaneous. The issue can be bypassed in K8s by either allowing more time for the container to start, or just by increasing allocated CPU resource (doubling it reduces container startup time to 60 seconds).

I think it'd be ideal to discuss whether this behaviour was expected and can we expect it to stay like this in future updates?

**Comment by thresheek** [comment]
Hi @eerotki, thanks for filing out this issue.

How many environment variables you have on your GKE cluster?  I suspect this is similar to what we observe in https://github.com/nginxinc/docker-nginx/issues/705 - can you please check if the https://github.com/nginxinc/docker-nginx/pull/706 would work for you (essentially add a prefix to env vars you're using in the configs) ?

Many thanks!

**Comment by eerotki** [comment]
Thanks for the reply @thresheek. There are 729 environment variables in total and as mentioned in https://github.com/nginxinc/docker-nginx/issues/705#issue-1374389618 this amount can be easily reached with GKE.

I tried the workaround suggested in https://github.com/nginxinc/docker-nginx/pull/706 by overriding the default `envsubst` script with new one, containing the changes for env filter. After this it takes 40 seconds for the service to start and is essentially the time it takes to process the envs with `printf '${%s} ' $(xargs -0n1 -a /proc/self/environ sh -c "echo \"\$@\" | grep -- \"${filter}\" | grep -oEm1 \"^[^=]+\"" --)`.

I mistakenly forgot to make my custom script executable at first, which resulted in `Ignoring, not executable`. By looking at the `envsubst` script, I'm starting to wonder whether there's any use for it if the `conf.d` is in read-only mode? Would it be better to first to check that and then try parse envs?

**Comment by eerotki** [comment]
@thresheek You are correct, my oversight on the order of script execution.

File system itself is writable. When testing locally with `-w "/etc/nginx/conf.d"` the folder is not writable in my configuration for user `nginx` that is used to run the container.
```
~ $ whoami
nginx
~ $ ls -la /etc/nginx/
total 48
dr-xr-xr-x    1 nginx    root          4096 Sep 19 00:20 .
drwxr-xr-x    1 root     root          4096 Sep 23 14:51 ..
dr-xr-xr-x    1 nginx    root          4096 Sep 23 15:15 conf.d
-r-xr-xr-x    1 nginx    root          1077 Jul 19 15:23 fastcgi.conf
-r-xr-xr-x    1 nginx    root          1007 Jul 19 15:23 fastcgi_params       
-r-xr-xr-x    1 nginx    root          5349 Jul 19 15:23 mime.types
lrwxrwxrwx    1 nginx    root            22 Sep 19 00:20 modules -> /usr/lib/nginx/modules
-r-xr-xr-x    1 nginx    root           826 Sep 19 00:20 nginx.conf
-r-xr-xr-x    1 nginx    root           636 Jul 19 15:23 scgi_params
-r-xr-xr-x    1 nginx    root           664 Jul 19 15:23 uwsgi_params
```
Why is the result not the same with the script?

**Comment by eerotki** [comment]
@thresheek I did some trial and error with `20-envsubst-on-templates.sh` script.

While the write check of `conf.d` is inside `auto_envsubst` function, the checking doesn't work as it is supposed to and no error is logged during start up.

```
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: info: can not modify /etc/nginx/conf.d/default.conf (read-only file system?)
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
/docker-entrypoint.sh: Configuration complete; ready for start up
2022/09/26 08:33:23 [notice] 1#1: using the "epoll" event method
2022/09/26 08:33:23 [notice] 1#1: nginx/1.23.1
2022/09/26 08:33:23 [notice] 1#1: built by gcc 11.2.1 20220219 (Alpine 11.2.1_git20220219)
...
```

However, if I modify the script by moving the write check outside of `auto_envsubst` function:
```
#!/bin/sh

...

output_dir="${NGINX_ENVSUBST_OUTPUT_DIR:-/etc/nginx/conf.d}"

...

if [ ! -w "$output_dir" ]; then
  entrypoint_log "$ME: ERROR: $output_dir is not writable"
  return 0
fi

auto_envsubst() {
  ...
}

auto_envsubst

exit 0
```
Then I correctly get an error on `20-envsubst-on-templates.sh`:
```
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: info: can not modify /etc/nginx/conf.d/default.conf (read-only file system?)
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
20-envsubst-on-templates.sh: ERROR: /etc/nginx/conf.d is not writable
/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
/docker-entrypoint.sh: Configuration complete; ready for start up
2022/09/26 08:33:52 [notice] 1#1: using the "epoll" event method
2022/09/26 08:33:52 [notice] 1#1: nginx/1.23.1
2022/09/26 08:33:52 [notice] 1#1: built by gcc 11.2.1 20220219 (Alpine 11.2.1_git20220219)
...
```
I'm curious whether you could reproduce this behaviour and figure out why there is a difference? I think this is a potential find to solve the issue.

**Comment by thresheek** [comment]
Do you have a template dir (`/etc/nginx/templates/`) mounted in your container?

**Comment by eerotki** [comment]
No, there's no `templates` folder.

**Comment by thresheek** [comment]
Thanks for such a detailed answer!

So it looks like reading from /proc/self/environ is slow, then, and filtering will not help in this case.

**Comment by thresheek** [comment]
Can you please test by using the following defined_envs:

```
defined_envs=$(printf '${%s} ' $(awk "END { for (name in ENVIRON) { print ( name ~ /${filter}/ ) ? name : \"\" } }" < /dev/null ))
```

Thank you!

**Comment by eerotki** [comment]
> Can you please test by using the following defined_envs:
> 
> ```
> defined_envs=$(printf '${%s} ' $(awk "END { for (name in ENVIRON) { print ( name ~ /${filter}/ ) ? name : \"\" } }" < /dev/null ))
> ```
> 
> Thank you!

Sure:
```
$ time defined_envs=$(printf '${%s} ' $(awk "END { for (name in ENVIRON) { print ( name ~ /${filter}/ ) ? name : \"\" } }" < /dev/null ))

real    0m0.097s
user    0m0.002s
sys     0m0.003s
```
732 envs filtered correctly to 14 and executed very quickly. Seems much better now! Tried with multi line variables as well and couldn't find issues. Would be great to see this merged.

**Comment by alessfg** [comment]
The fix has been merged as of [0029c30](https://github.com/nginxinc/docker-nginx-unprivileged/commit/0029c30fecb7aa34e4b22e1b9adac4a2ebbd492a)! Hope it works out 😄
